### PR TITLE
feat: add docker resource detector

### PR DESCRIPTION
**Pull request by abhee11**
## Which problem is this PR solving?

- [abstracting the docker container id logic ](https://github.com/open-telemetry/opentelemetry-js-contrib/issues/907)

## Short description of the changes

- adding a new detector for docker container id cgroupov1 only 

## Checklist

- [ ] Ran `npm run test-all-versions` for the edited package(s) on the latest commit if applicable.

### CODE REVIEW

File: detectors/node/opentelemetry-resource-detector-docker/package.json
Comment (vmarchaud) [inline detectors/node/opentelemetry-resource-detector-docker/package.json:56]: this should be a dev dependencies

File: detectors/node/opentelemetry-resource-detector-docker/package.json
Comment (vmarchaud) [inline detectors/node/opentelemetry-resource-detector-docker/package.json:3]: not sure we can release directly as 1.0 ? waiting for other maintainers input on this

File: detectors/node/opentelemetry-resource-detector-docker/package.json
Comment (vmarchaud) [inline detectors/node/opentelemetry-resource-detector-docker/package.json:53]: i believe this isn't used

**Review by vmarchaud — COMMENTED** [review]
the detector itself looks good to me, you'll need to fix the CLA though

File: detectors/node/opentelemetry-resource-detector-docker/package.json
Comment (abhee11) [reply detectors/node/opentelemetry-resource-detector-docker/package.json:53]: done

File: detectors/node/opentelemetry-resource-detector-docker/package.json
Comment (abhee11) [reply detectors/node/opentelemetry-resource-detector-docker/package.json:56]: makes sense - done

**Review by vmarchaud — COMMENTED** [review]
you'll also need to add there: https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/.release-please-manifest.json so it can be automatically published

**Comment by dyladan** [comment]
> you'll also need to add there: [https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/.release-please-manifest.json](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/.release-please-manifest.json?rgh-link-date=2022-03-07T17%3A52%3A36Z) so it can be automatically published

I think this isn't true anymore with release-please v3

**Comment by abhee11** [comment]
Just cleaning out some lint issues - waiting for jobs to run - @dyladan @vmarchaud 
I updated the .release-please-manifest.json as per the comment - let me know if you like me to revert it.

I appreciate the review on this - thank you

File: detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts
Comment (Rauno56) [reply detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts:39]: [fs docs](https://nodejs.org/api/fs.html#fspromisesaccesspath-mode) seem to agree:

> Using fsPromises.access() to check for the accessibility of a file before calling fsPromises.open() is not recommended. Doing so introduces a race condition, since other processes may change the file's state between the two calls. Instead, user code should open/read/write the file directly and handle the error raised if the file is not accessible.

Comment (abhee11) [reply detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts:39]: @Rauno56 That is a good insight , thank you for the comment. I will update the PR and notify .

Comment (abhee11) [reply detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts:39]: So - I will remove the check itself and if there is any issue - it should be caught in the catch. Right ?

Comment (Rauno56) [reply detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts:39]: Yes - The catch surrounding `readFile`.

Comment (abhee11) [reply detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts]: done

**Comment by abhee11** [comment]
@dyladan @Rauno56 Seems like lint is failing - I tested my local with "lerna run lint" and it fails on the "aws" package - but in the logs it says it fails on "docker" package - not able to replicate ? any advice ?


npm run lint runs fine on my local

**Comment by vmarchaud** [comment]
> I think this isn't true anymore with release-please v3

Didn't saw that sorry @abhee11

**Comment by Rauno56** [comment]
> @dyladan @Rauno56 Seems like lint is failing - I tested my local with "lerna run lint" and it fails on the "aws" package - but in the logs it says it fails on "docker" package - not able to replicate ? any advice ?
> 
> npm run lint runs fine on my local

Messages you are referring to(from aws) are only warnings and do not cause the failure. The relevant part for you is this:

```
> @opentelemetry/resource-detector-docker@1.0.0 lint /__w/opentelemetry-js-contrib/opentelemetry-js-contrib/detectors/node/opentelemetry-resource-detector-docker
> eslint . --ext .ts


lerna ERR! npm run lint stderr:

Oops! Something went wrong! :(

ESLint: 8.7.0

ESLint couldn't find a configuration file. To set up a configuration file for this project, please run:

    npm init @eslint/config

ESLint looked for configuration files in /__w/opentelemetry-js-contrib/opentelemetry-js-contrib/detectors/node/opentelemetry-resource-detector-docker/src/detectors and its ancestors. If it found none, it then looked in your home directory.

If you think you already have a configuration file or if you need more help, please stop by the ESLint chat room: https://eslint.org/chat/help
```

Make sure all the relevant files are checked in and paths in them specified correctly.

**Comment by abhee11** [comment]
@Rauno56 Thanks - I was missing the eslintrc and ignore files - added them.

**Review by Rauno56 — APPROVED** [review]

File: detectors/node/opentelemetry-resource-detector-docker/.eslintignore
Comment (dyladan) [inline detectors/node/opentelemetry-resource-detector-docker/.eslintignore:1]: nit: add newline at the end

File: detectors/node/opentelemetry-resource-detector-docker/.npmignore
Comment (dyladan) [inline detectors/node/opentelemetry-resource-detector-docker/.npmignore:4]: nit: newline

File: detectors/node/opentelemetry-resource-detector-docker/package.json
Comment (dyladan) [reply detectors/node/opentelemetry-resource-detector-docker/package.json:3]: I would definitely prefer to start with 0.1.0

File: detectors/node/opentelemetry-resource-detector-docker/package.json
Comment (dyladan) [inline detectors/node/opentelemetry-resource-detector-docker/package.json:4]: There may be other things added in the future. I would make this more general
```suggestion
  "description": "OpenTelemetry resource detector for docker resource attributes",
```

File: detectors/node/opentelemetry-resource-detector-docker/README.md
Comment (dyladan) [inline detectors/node/opentelemetry-resource-detector-docker/README.md:32]: A detector per attribute? Seems like it would be better to have a `DockerDetector` and list the attributes it detects here

File: detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts
Comment (dyladan) [inline detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts:56]: is it always \n or can it be \r\n?
```suggestion
      const splitData = rawData.trim().split(/\r?\n/);
```

**Review by dyladan — CHANGES_REQUESTED** [review]
Please use 0.1.0 as the initial version. We can bump it to 1.0.0 when we know it is stable. Blocking for the 1.0.0 issue.

**Comment by Flarna** [comment]
I guess release-please-config.json needs to be adapted to get this published after merge.

File: detectors/node/opentelemetry-resource-detector-docker/README.md
Comment (abhee11) [reply detectors/node/opentelemetry-resource-detector-docker/README.md:32]: @dyladan thank you for the comments
Currently I have named it as DockerCGroupV1Detector given it is specific to CGroup v1 
We currently just have container ID that is added as a part of resouce, In future we could definitely add more attributes but I think it is important to mark the name CGroupV1 

Do you agree ?

File: detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts
Comment (abhee11) [reply detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts:56]: Because docker is linux based we should be okay with \n

File: detectors/node/opentelemetry-resource-detector-docker/.npmignore
Comment (abhee11) [reply detectors/node/opentelemetry-resource-detector-docker/.npmignore:4]: done

File: detectors/node/opentelemetry-resource-detector-docker/package.json
Comment (abhee11) [reply detectors/node/opentelemetry-resource-detector-docker/package.json:4]: makes sense - done

**Comment by abhee11** [comment]
@Flarna As per comments above - I believe we dont need that anymore ?
@dyladan Please review changes and my comments for \r and the READ.ME change - let me know what you think

**Comment by Flarna** [comment]
@abhee11 no need anymore to touch `.release-please-manifest.json` but still needed to add a new package in `release-please-config.json`.

See https://github.com/open-telemetry/opentelemetry-js-contrib/pull/938 which shows that this is still needed.

**Comment by abhee11** [comment]
@Flarna my bad - pushed - thanks

File: detectors/node/opentelemetry-resource-detector-docker/README.md
Comment (dyladan) [reply detectors/node/opentelemetry-resource-detector-docker/README.md:32]: OK yeah thats fine

**Review by dyladan — APPROVED** [review]

**Review by vmarchaud — APPROVED** [review]

File: detectors/node/opentelemetry-resource-detector-docker/README.md
Comment (blumamir) [inline detectors/node/opentelemetry-resource-detector-docker/README.md:10]: According to [the semantic conventions for resource container](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/container.md), the following 5 attributes belong to the `docker` resource:
```
container.name, container.id, container.runtime, container.image.name, container.image.tag
```

The current implementation only extracts a single value (`container.id`) of these 5 and only for `CGroupV1` (not sure how many other types are possible).

I suggest to make that clear in the README, or support the other fields in the spec

File: detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts
Comment (blumamir) [inline detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts:31]: If I'm not mistaken, this is not used anywhere

File: detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts
Comment (abhee11) [reply detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts:45]: So - the reason for this is - if you are installing DockerResourceV1Detector - that would mean - you are adding this dependency assuming you have v1 version running - hence the warning !

As this dependency would be of no use if it did get them the container id - hence we need to warn them that they need have a diff docker version running.

File: detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts
Comment (abhee11) [reply detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts:58]: Yes - for group v1 - this seems to be the implementation across java , go etc - hence I adapted the same for JS

File: detectors/node/opentelemetry-resource-detector-docker/README.md
Comment (abhee11) [reply detectors/node/opentelemetry-resource-detector-docker/README.md:10]: currently in the readme it does say that we populate only container id - I will add more details in case if it is missing - thanks

File: detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts
Comment (abhee11) [reply detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts:68]: We need a way to let the user know that their dockerV1Detector is not doing as they hoped. I am wondering if their a another way - do you have any suggestions ?

File: detectors/node/opentelemetry-resource-detector-docker/test/DockerCGroupv1Detector.test.ts
Comment (blumamir) [inline detectors/node/opentelemetry-resource-detector-docker/test/DockerCGroupv1Detector.test.ts:38]: nit: `emoty` -> `empty` :)

File: detectors/node/opentelemetry-resource-detector-docker/test/DockerCGroupv1Detector.test.ts
Comment (blumamir) [reply detectors/node/opentelemetry-resource-detector-docker/test/DockerCGroupv1Detector.test.ts:79]: Maybe I misunderstand you, but the test uses:
```js
      const errorMsg = {
        fileNotFoundError: new Error('cannot file in the path}'),
      };
```

And it will not return an error, but an object with one key named `fileNotFoundError` which is an `Error`.
Is it consistent with what `fs` returns on `err`?
I would expect it to be something like:
```js
.rejects(new Error('foo bar'));
```

File: detectors/node/opentelemetry-resource-detector-docker/test/DockerCGroupv1Detector.test.ts
Comment (abhee11) [reply detectors/node/opentelemetry-resource-detector-docker/test/DockerCGroupv1Detector.test.ts:39]: Sorry - I don't understand this comment - can you explain a little more
It should return { attributes: {} } as resource if we do not give it the right file path ( in case of test - the right stub which resolves into the container id  ) ?

File: detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts
Comment (blumamir) [reply detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts]: I am wondering if including a resource detector means that the user is expecting that resource to be present. 

2 Examples: 
1. If the user installs an aws detector and then runs his service locally, I wouldn't expect any warnings to be printed to log as this is completely valid and the message will be confusing. Maybe the service is running inside a docker in production but not in other environments.
2. If someone creates a distribution of OpenTelemetry and includes many detectors, the expectation (as far as I am aware) is that only the relevant detectors will be applied. Detectors that are not relevant should be silently ignored. This is the way instrumentations are handled: if someone installs the `@opentelemetry/auto-instrumentations-node` package with all instrumentations, and then his app does not require `redis` package, then `instrumentation-redis` will just not load with no warning to diag channel

I think emitting it as `info` makes more sense, something like: "Docker CGROUP V1 Detector did not identify running inside a docker container, no docker attributes will be added to resource".

This is for the case that the file does not exist. If the file exists but exception is thrown for any other reason, I would expect an `error` with message that includes the `errorMessage`.

This is my take on it, but I'm open for other opinions

File: detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts
Comment (blumamir) [reply detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts:44]: See my comment on the other thread

File: detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts
Comment (blumamir) [reply detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts]: Quoting from the `DiagLogger` interface:

```js
  /**
   * Log a warning scenario to inform the developer of an issues that should be investigated.
   * The requested operation may or may not have succeeded or completed.
   */
  warn: DiagLogFunction;
```

I think when the file is missing it doesn't mean that the developer should investigate anything but only be informed

File: detectors/node/opentelemetry-resource-detector-docker/test/DockerCGroupv1Detector.test.ts
Comment (blumamir) [reply detectors/node/opentelemetry-resource-detector-docker/test/DockerCGroupv1Detector.test.ts:39]: Maybe it's not really a problem 🤷‍♂️, but I can't understand why the invocation of this detector does not return non-empty resource when run in CI.

Comment (abhee11) [reply detectors/node/opentelemetry-resource-detector-docker/test/DockerCGroupv1Detector.test.ts:43]: I have added a stub to return undefined - just to be more careful

File: detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts
Comment (abhee11) [reply detectors/node/opentelemetry-resource-detector-docker/src/detectors/DockerCGroupV1Detector.ts]: This is very helpful - thank you for taking the time and helping me out with this. I appreciate it. 
Addressed.

**Comment by abhee11** [comment]
@blumamir I have addressed your comments and pushed the changes. Let me know if there is anything else.

File: detectors/node/opentelemetry-resource-detector-docker/test/DockerCGroupv1Detector.test.ts
Comment (dyladan) [reply detectors/node/opentelemetry-resource-detector-docker/test/DockerCGroupv1Detector.test.ts:43]: Does the resource detector require some sort of elevated permission or security context? I'm also curious why this returns undefined in CI because our CI runs inside docker containers.

Comment (abhee11) [reply detectors/node/opentelemetry-resource-detector-docker/test/DockerCGroupv1Detector.test.ts:43]: what version of docker is it running on ? if it is v2 - then this is expected to return undefined as the cgroup path is different and may need permission.

**Comment by abhee11** [comment]
@blumamir @dyladan Let me know if there is anything more I need to refactor - I have made the proposed changes

**Review by blumamir — APPROVED** [review]
LGTM. thanks for the patient and for addressing everything

**Comment by abhee11** [comment]
@blumamir Thank you, feel free to merge.

**Pull request merged**
